### PR TITLE
add listener registration to Store.

### PR DESCRIPTION
**Pull request by RafeArnold**
Adds the ability to register listeners to a Store. Only implemented for InMemoryObjectStore for now. Default implementation throws an unsupported exception.

## Submitter checklist

- [x] Recommended: Join [WireMock Slack](https://slack.wiremock.org/) to get any help in `#help-contributing` or a project-specific channel like `#wiremock-java`
- [x] The PR request is well described and justified, including the body and the references
- [x] The PR title represents the desired changelog entry
- [x] The repository's code style is followed (see the contributing guide)
- [x] Test coverage that demonstrates that the change works as expected
- [ ] For new features, there's necessary documentation in this pull request or in a subsequent PR to [wiremock.org](https://github.com/wiremock/wiremock.org)

### CODE REVIEW

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java [hunk 1]
```diff
@@ -27,6 +32,7 @@ public class InMemoryObjectStore implements ObjectStore {
   private final ConcurrentHashMap<String, Object> cache;
   private final Queue<String> keyUseOrder = new ConcurrentLinkedQueue<>();
   private final int maxItems;
+  private final List<StoreEventHandler<String, Object>> listeners = new ArrayList<>();
```
Comment (tomakehurst) [inline src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java:35]: Any reason we can't just use `Consumer<StoreEvent>` as the listener type?

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java [hunk 2]
```diff
@@ -54,27 +60,39 @@ public Stream<String> getAllKeys() {
...
+                (k, currentValue) -> {
+                  previousValue.set((T) currentValue);
+                  return valueFunction.apply((T) currentValue);
+                });
     if (result != null) {
       touchAndResize(key);
     } else {
       keyUseOrder.remove(key);
     }
+    handleEvent(new StoreEvent<>(key, previousValue.get(), result));
```
Comment (tomakehurst) [inline src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java:85]: Might be nicer to have two different static factory methods i.e.
`StoreEvent.set(key, prevValue, newValue()` and `StoreEvent.remove(key, prevValue)`

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java (hunk 1 above)
Comment (RafeArnold) [reply src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java:35]: yep sounds good to me

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java (hunk 2 above)
Comment (RafeArnold) [reply src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java:85]: yeah that's nice

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java (hunk 1 above)
Comment (RafeArnold) [reply src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java]: done

File: src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java (hunk 2 above)
Comment (RafeArnold) [reply src/main/java/com/github/tomakehurst/wiremock/store/InMemoryObjectStore.java]: done

**Pull request merged**